app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx

```diff
@@ -54,284 +73,310 @@ const replaceTextArea = ({
+        return;
       }
-    };
-
-    document.addEventListener('click', closeOnClickOutsideListener);
 
-    return () =>
-      document.removeEventListener('click', closeOnClickOutsideListener);
-  }, [searchTerm]);
+      if (cachedSearches[searchTerm]) {
+        setUsers(cachedSearches[searchTerm]);
+        return;
+      }
 
-  useLayoutEffect(() => {
-    const { current: plainTextInput } = plainTextAreaRef;
-    const { current: combobox } = comboboxRef;
+      fetchSuggestions(searchTerm).then(({ result: fetchedUsers }) => {
+        const resultLength = Math.min(
+          fetchedUsers.length,
+          MAX_RESULTS_DISPLAYED,
+        );
 
-    const isComboboxVisible = !combobox.classList.contains('hidden');
-    const activeInput = isComboboxVisible ? combobox : plainTextInput;
-    activeInput.focus();
-    activeInput.setSelectionRange(cursorPosition, cursorPosition - 1);
-  }, [cursorPosition]);
+        const results = fetchedUsers.slice(0, resultLength);
 
-  const handleTextInputChange = ({ target: { value } }) => {
-    setTextContent(value);
-    const isComboboxVisible = !comboboxRef.current.classList.contains('hidden');
-    const currentActiveInput = isComboboxVisible
-      ? comboboxRef.current
-      : plainTextAreaRef.current;
+        setCachedSearches({
+          ...cachedSearches,
+          [searchTerm]: results,
+        });
 
-    const { isUserMention, indexOfMentionStart } = getMentionWordData(
-      currentActiveInput,
-    );
+        setUsers(results);
 
-    const { selectionStart } = currentActiveInput;
+        // Let screen reader users know a list has populated
+        if (!ariaHelperText && fetchedUsers.length > 0) {
+          setAriaHelperText(
+            `Mention user, ${fetchedUsers.length} results found`,
+          );
+        }
+      });
+    }, [searchTerm, fetchSuggestions, cachedSearches, ariaHelperText]);
 
-    if (isUserMention) {
-      // search term begins after the @ character
-      const searchTermStartPosition = indexOfMentionStart + 1;
+    useLayoutEffect(() => {
+      const popover = popoverRef.current;
+      if (!popover) {
+        return;
+      }
 
-      const mentionText = value.substring(
-        searchTermStartPosition,
-        selectionStart,
+      const closeOnClickOutsideListener = (event) => {
+        if (!popover.contains(event.target)) {
+          // User clicked outside, reset to not searching state
+          setSearchTerm('');
+          setAriaHelperText('');
+          setUsers([]);
+        }
+      };
+
+      document.addEventListener('click', closeOnClickOutsideListener);
+
+      return () =>
+        document.removeEventListener('click', closeOnClickOutsideListener);
+    }, [searchTerm]);
+
+    useLayoutEffect(() => {
+      const { current: plainTextInput } = plainTextAreaRef;
+      const { current: combobox } = comboboxRef;
+
+      const isComboboxVisible = !combobox.classList.contains('hidden');
```

Comment:
Nitpick (non-blocking): `isComboboxVisible` is only used right below where it's defined. Prefer inlining the condition.

Reply:
it started life that way,  but I thought maybe it's better to be clearer what we're checking? I dunno I think sometimes I veer on the side of creating a named variable just for readability (maybe too often 🙃)